### PR TITLE
Fixes #20496 - Add API for canceling jobs

### PR DESCRIPTION
**Pull request by adamruzicka**
- [x] tests

### CODE REVIEW

File: app/controllers/api/v2/job_invocations_controller.rb [hunk 1]
```diff
@@ -85,12 +85,32 @@ def output
         end
       end
 
+      api :POST, '/job_invocations/:id/cancel'
+      param :id, :identifier, :required => true
+      param :force, :bool
```
Comment (iNecas) [inline app/controllers/api/v2/job_invocations_controller.rb]: Please add a description of what that means.

File: app/controllers/api/v2/job_invocations_controller.rb [hunk 2]
```diff
@@ -85,12 +85,32 @@ def output
...
+      end
+
       private
 
       def action_permission
         case params[:action]
         when 'output'
           :view
+        when 'cancel'
+          :view
```
Comment (iNecas) [inline app/controllers/api/v2/job_invocations_controller.rb]: Should this be `:view`?

Comment (adamruzicka) [reply app/controllers/api/v2/job_invocations_controller.rb]: The user needs to be able to view the job invocation and edit tasks. I can create a dedicated permission for editing job invocations if needed, so far we don't have one.

Comment (iNecas) [reply app/controllers/api/v2/job_invocations_controller.rb]: +1 for addiing it.

**Comment by iNecas** [comment]
Tests are failing

**Comment by adamruzicka** [comment]
The tests are finally green

File: lib/foreman_remote_execution/engine.rb [hunk 3]
```diff
@@ -56,6 +56,7 @@ class Engine < ::Rails::Engine
                                                 'api/v2/job_invocations' => [:create] }, :resource_type => 'JobInvocation'
           permission :view_job_invocations, { :job_invocations => [:index, :show, :auto_complete_search], :template_invocations => [:show],
                                               'api/v2/job_invocations' => [:index, :show, :output] }, :resource_type => 'JobInvocation'
+          permission :cancel_job_invocations, { 'api/v2/job_invocations' => [:cancel] }, :resource_type => 'JobInvocation'
```
Comment (iNecas) [inline lib/foreman_remote_execution/engine.rb]: You're adding `:cancel_job_invocation` here, while adding `:edit_job_invocations` permission later to the `MANAGER_PERMISSIONS`.  I would probably stick with `:edit_job_invocations` here

Comment (adamruzicka) [reply lib/foreman_remote_execution/engine.rb]: There is a double-save happening when a job invocation is being created by the composer (probably due to associations). This means a user has to have both `:create` and `:edit` permissions to be able to run jobs. I would like to go with `:cancel` before we get this sorted out

Comment (iNecas) [reply lib/foreman_remote_execution/engine.rb]: I see, we should then unify the way how the UI and API uses the permission, by using `:cancel_job_invocations` for the UI as well

File: app/views/template_invocations/show.html.erb [hunk 4]
```diff
@@ -9,7 +9,7 @@
                      link_to_function(_('Toggle STDERR'), '$("div.line.stderr").toggle()', :class => 'btn btn-default'),
                      link_to_function(_('Toggle STDOUT'), '$("div.line.stdout").toggle()', :class => 'btn btn-default'),
                      link_to_function(_('Toggle DEBUG'), '$("div.line.debug").toggle()', :class => 'btn btn-default')) %>
-    <%= button_group(template_invocation_task_buttons(@template_invocation_task)) %>
+    <%= button_group(template_invocation_task_buttons(@template_invocation_task, @template_invocation.job_invocation)) %>
```
Comment (iNecas) [inline app/views/template_invocations/show.html.erb]: I don't see the `job_invocation_task_buttons` to have updated the list of arguments. Am I missing something here, or this PR is? :)

Comment (iNecas) [reply app/views/template_invocations/show.html.erb]: Taking back: I was looking at wrong method. NM

File: app/helpers/remote_execution_helper.rb [hunk 5]
```diff
@@ -116,7 +116,7 @@ def job_invocation_task_buttons(task)
                          :class => 'btn btn-default',
                          :title => _('See the last task details'))
     end
-    if authorized_for(:permission => :edit_foreman_tasks, :auth_object => task, :authorizer => task_authorizer)
+    if authorized_for(:permission => :cancel_job_invocations, :auth_object => job_invocation)
       buttons << link_to(_('Cancel Job'), cancel_foreman_tasks_task_path(task),
```
Comment (iNecas) [inline app/helpers/remote_execution_helper.rb]: It doesn't work as expected: I had an user that had only rex manager rigths: I got an error

```
Please request one of the required permissions listed below from a Foreman administrator:

    edit_foreman_tasks

```
We probably need a separate end-point for the job cancellation, and not use the task end point for it.

**Comment by adamruzicka** [comment]
@iNecas could you please take another look?

File: app/controllers/job_invocations_controller.rb [hunk 6]
```diff
@@ -72,6 +72,26 @@ def preview_hosts
...
+                         _('Trying to cancel the job')
+                       end
+    else
+      flash[:warning] = if params[:force]
+                          _('The job cannot be aborted at the moment.')
+                        else
+                          _('The job cannot be cancelled at the moment.')
+                        end
+    end
+    redirect_to :back
```
Comment (iNecas) [inline app/controllers/job_invocations_controller.rb]: This doesn't work in rails 5, see https://github.com/theforeman/foreman-tasks/pull/308/files on how to transition that

**Comment by iNecas** [comment]
Works as expected. The tests were passing before and I don't see the latest updates to cause any additional failure. Merging now

**Pull request merged**